Speed up `TokenCursor`

nnethercote opened the pull request:
Plus a few related clean-ups.

r? @petrochenkov

nnethercote commented:
Some local instruction count measurements:

Benchmark | Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | -- | --
time-macros-0.2.3 | check | full | -4.31% | 21.54x
async-std-1.10.0 | check | full | -4.17% | 20.85x
tt-muncher | check | full | -3.34% | 16.68x
yansi-0.5.0 | check | full | -2.16% | 10.80x
quote-stress | check | full | -2.11% | 10.57x
web-sys-0.3.56 | check | full | -1.09% | 5.46x
num-derive-0.3.3 | check | full | -1.09% | 5.45x
ctor-0.1.21 | check | full | -1.05% | 5.27x
futures-macro-0.3.19 | check | full | -0.95% | 4.76x
mockall_derive-0.11.0 | check | full | -0.89% | 4.47x
wasm-bindgen-backend-0.2.79 | check | full | -0.88% | 4.42x
tonic-build-0.7.0 | check | full | -0.87% | 4.33x
vsdb_derive-0.21.1 | check | full | -0.84% | 4.20x
pest_generator-2.1.3 | check | full | -0.83% | 4.13x
enum-as-inner-0.4.0 | check | full | -0.81% | 4.06x
stdweb-derive-0.5.3 | check | full | -0.81% | 4.03x
scroll_derive-0.11.0 | check | full | -0.80% | 4.02x
serde_derive-1.0.136 | check | full | -0.75% | 3.73x
clap_derive-3.1.7 | check | full | -0.73% | 3.67x
ref-cast-impl-1.0.6 | check | full | -0.73% | 3.63x
wayland-scanner-0.29.4 | check | full | -0.71% | 3.55x
prost-derive-0.10.0 | check | full | -0.69% | 3.47x
pyo3-macros-backend-0.16.3 | check | full | -0.69% | 3.46x
diesel_derives-1.4.1 | check | full | -0.67% | 3.36x
inotify-0.10.0 | check | full | -0.66% | 3.32x
funty-2.0.0 | check | full | -0.63% | 3.13x
enumflags2_derive-0.7.4 | check | full | -0.61% | 3.06x
deep-vector | check | full | -0.48% | 2.40x
nix-0.23.1 | check | full | -0.47% | 2.33x
futures-lite-1.12.0 | check | full | -0.33% | 1.66x
externs | check | full | -0.31% | 1.55x

The cycles counts are even better, e.g.:



Benchmark | Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | -- | --
tt-muncher | check | full | -9.67% | 48.36x
async-std-1.10.0 | check | full | -8.13% | 40.67x
quote-stress | check | full | -7.13% | 35.65x
time-macros-0.2.3 | check | full | -6.77% | 33.85x
token-stream-stress | check | full | -3.95% | 19.75x
yansi-0.5.0 | check | full | -3.74% | 18.72x
web-sys-0.3.56 | check | full | -3.50% | 17.48x
tokio-1.17.0 | check | full | -3.22% | 16.11x
log-0.4.16 | check | full | -3.20% | 15.98x
futures-macro-0.3.19 | check | full | -3.18% | 15.92x
inotify-0.10.0 | check | full | -3.14% | 15.69x
issue-88862 | check | full | 2.98% | 14.89x
externs | check | full | -2.62% | 13.08x
unify-linearly | check | full | -2.47% | 12.36x
enumflags2_derive-0.7.4 | check | full | -2.14% | 10.69x
mockall_derive-0.11.0 | check | full | -2.09% | 10.43x
ref-cast-impl-1.0.6 | check | full | -1.98% | 9.88x
tonic-build-0.7.0 | check | full | -1.92% | 9.59x
scroll_derive-0.11.0 | check | full | -1.86% | 9.31x
stdweb-derive-0.5.3 | check | full | -1.76% | 8.79x

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 36c7c829e99b3785de94e6113b4abb4ff0410e9c with merge efe204aef96605ab9df88a21434b2b571a770ede...
<!-- homu: {"type":"TryBuildStarted","head_sha":"36c7c829e99b3785de94e6113b4abb4ff0410e9c","merge_sha":"efe204aef96605ab9df88a21434b2b571a770ede"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6076821531?check_suite_focus=true)
Build commit: efe204aef96605ab9df88a21434b2b571a770ede (`efe204aef96605ab9df88a21434b2b571a770ede`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6076821531?check_suite_focus=true"},"merge_sha":"efe204aef96605ab9df88a21434b2b571a770ede"} -->

rust-timer commented:
Queued efe204aef96605ab9df88a21434b2b571a770ede with parent e2661bac6d8ae6b9e2a26cd08e9c087ab59e1357, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e2661bac6d8ae6b9e2a26cd08e9c087ab59e1357&end=efe204aef96605ab9df88a21434b2b571a770ede).

rust-timer commented:
Finished benchmarking commit (efe204aef96605ab9df88a21434b2b571a770ede): [comparison url](https://perf.rust-lang.org/compare.html?start=e2661bac6d8ae6b9e2a26cd08e9c087ab59e1357&end=efe204aef96605ab9df88a21434b2b571a770ede).

**Summary**: 
- Primary benchmarks: 🎉 relevant improvements found
- Secondary benchmarks: 🎉 relevant improvements found

| | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:---:|:---:|:---:|:---:|:---:|:---:|
| count[^1] | 0 | 0 | 22 | 20 | 22 |
| mean[^2] | N/A | N/A | -0.3% | -1.5% | -0.3% |
| max | N/A | N/A | -0.6% | -4.1% | -0.6% |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

nnethercote commented:
`tt-muncher` is most affected. The instruction count results for it are good, with 1-4% reductions. But the wall-time results are amazing, with 6-16% reductions. My guess would be this is due to fewer branch mispredictions.

nnethercote commented:
I have fixed the tidy error.

Cachegrind indicates that the outsized wall-time speedup for `tt-muncher` is actually due to fewer loads and stores in and around `Parser::bump()`, though it was hard to tell exactly which part of the change was responsible.

nnethercote commented:
I made a bunch more improvements. Instruction counts:

Benchmark | Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | -- | --
async-std-1.10.0 | check | full | -14.58% | 72.91x
time-macros-0.2.3 | check | full | -13.40% | 67.00x
tt-muncher | check | full | -12.18% | 60.91x
yansi-0.5.0 | check | full | -7.51% | 37.54x
ctor-0.1.21 | check | full | -3.57% | 17.86x
num-derive-0.3.3 | check | full | -3.55% | 17.77x
web-sys-0.3.56 | check | full | -3.38% | 16.91x
futures-macro-0.3.19 | check | full | -3.11% | 15.55x
mockall_derive-0.11.0 | check | full | -2.96% | 14.80x
wasm-bindgen-backend-0.2.79 | check | full | -2.95% | 14.76x
tonic-build-0.7.0 | check | full | -2.94% | 14.72x
vsdb_derive-0.21.1 | check | full | -2.90% | 14.52x
scroll_derive-0.11.0 | check | full | -2.89% | 14.47x
pest_generator-2.1.3 | check | full | -2.83% | 14.14x
enum-as-inner-0.4.0 | check | full | -2.76% | 13.79x
stdweb-derive-0.5.3 | check | full | -2.75% | 13.73x
serde_derive-1.0.136 | check | full | -2.58% | 12.89x
clap_derive-3.1.7 | check | full | -2.57% | 12.84x
prost-derive-0.10.0 | check | full | -2.47% | 12.34x
wayland-scanner-0.29.4 | check | full | -2.45% | 12.27x
ref-cast-impl-1.0.6 | check | full | -2.45% | 12.27x
pyo3-macros-backend-0.16.3 | check | full | -2.40% | 12.02x
diesel_derives-1.4.1 | check | full | -2.33% | 11.63x
inotify-0.10.0 | check | full | -2.19% | 10.94x
funty-2.0.0 | check | full | -2.10% | 10.50x
enumflags2_derive-0.7.4 | check | full | -2.09% | 10.43x

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit e5d0a36920656874a7ad1251c6f40ece3d2164aa with merge 90d9443d0fb0f81ed8a2e2abcd0ae2f64af31b5d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e5d0a36920656874a7ad1251c6f40ece3d2164aa","merge_sha":"90d9443d0fb0f81ed8a2e2abcd0ae2f64af31b5d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6092380205?check_suite_focus=true)
Build commit: 90d9443d0fb0f81ed8a2e2abcd0ae2f64af31b5d (`90d9443d0fb0f81ed8a2e2abcd0ae2f64af31b5d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6092380205?check_suite_focus=true"},"merge_sha":"90d9443d0fb0f81ed8a2e2abcd0ae2f64af31b5d"} -->

rust-timer commented:
Finished benchmarking commit (90d9443d0fb0f81ed8a2e2abcd0ae2f64af31b5d): [comparison url](https://perf.rust-lang.org/compare.html?start=27af5175497936ea3413bef5816e7c0172514b9c&end=90d9443d0fb0f81ed8a2e2abcd0ae2f64af31b5d).

**Summary**: 
- Primary benchmarks: 🎉 relevant improvements found
- Secondary benchmarks: 🎉 relevant improvements found

| | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:---:|:---:|:---:|:---:|:---:|:---:|
| count[^1] | 0 | 0 | 73 | 52 | 73 |
| mean[^2] | N/A | N/A | -0.7% | -2.4% | -0.7% |
| max | N/A | N/A | -3.2% | -12.8% | -3.2% |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

petrochenkov commented on compiler/rustc_parse/src/parser/mod.rs:
```diff
@@ -347,7 +347,7 @@ impl TokenCursor {
                         .collect::<TokenStream>()
                 },
-                false,
+                true,
```
Any assumptions like this make fixing https://github.com/rust-lang/rust/issues/67062 harder.

We should drop `NoDelim` delimiters as late as possible, in a single place, and with a large FIXME.

petrochenkov replied:
https://github.com/rust-lang/rust/pull/96210#discussion_r854037894

petrochenkov commented on compiler/rustc_parse/src/parser/mod.rs:
```diff
@@ -984,12 +984,6 @@ impl<'a> Parser<'a> {
-        if self.prev_token.kind == TokenKind::Eof {
-            let msg = "attempted to bump the parser past EOF (may be stuck in a loop)";
-            self.span_bug(self.token.span, msg);
-        }
```
We 100% have seen this error, and relatively commonly, until relevant bugs in the parser were fixed.
But it was several years ago already.

petrochenkov commented on compiler/rustc_parse/src/parser/mod.rs:
```diff
@@ -994,22 +994,22 @@ impl<'a> Parser<'a> {
         // an unglued token. See `break_and_eat` for more details
         self.token_cursor.break_last_token = false;
-        if next.span.is_dummy() {
+        if next.0.span.is_dummy() {
```
It's awful that it has to be done for performance.
I pretty much never use `tuple.0` indexing and immediately destructure for readability.
I remember asking Aaron1011 to use destructuring on this specific line during review, when he originally added the spacing treatment to this code.

nnethercote replied:
I agree. I tried changing a couple of other hot functions (`inlined_next`, `inlined_bump_with`) that use the same kind of destructuring, but there was no performance change, so I left them alone.

nnethercote replied:
I'll add a comment explaining this.

petrochenkov commented on compiler/rustc_parse/src/parser/mod.rs:
```diff
@@ -238,73 +241,64 @@ struct TokenCursor {
-            open_delim: false,
             tree_cursor: tts.into_trees(),
-            close_delim: false,
+            need_to_produce_close_delim: delim != DelimToken::NoDelim,
```
```suggestion
            need_to_produce_close_delim: true,
```
Can all the `NoDelim` treatment for the closing token be moved to `fn inlined_next`, in a similar way to the opening token?
Then it would be the place to add the FIXME about https://github.com/rust-lang/rust/issues/67062.

nnethercote replied:
Ok, I will do this.

nnethercote commented:
I have added two new commits. The first changes how `CloseDelim` is handled, making things neater, and also putting all the `NoDelim` handling within `inlined_next`. The second avoids some cloning for more performance improvements. Updated instruction counts:

Benchmark | Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | -- | --
async-std-1.10.0 | check | full | -18.07% | 90.33x
time-macros-0.2.3 | check | full | -16.77% | 83.87x
tt-muncher | check | full | -15.11% | 75.53x
yansi-0.5.0 | check | full | -9.16% | 45.82x
ctor-0.1.21 | check | full | -4.48% | 22.42x
num-derive-0.3.3 | check | full | -4.47% | 22.33x
web-sys-0.3.56 | check | full | -4.43% | 22.14x
futures-macro-0.3.19 | check | full | -3.90% | 19.48x
tonic-build-0.7.0 | check | full | -3.70% | 18.51x
mockall_derive-0.11.0 | check | full | -3.69% | 18.46x
wasm-bindgen-backend-0.2.79 | check | full | -3.68% | 18.42x
vsdb_derive-0.21.1 | check | full | -3.61% | 18.03x
scroll_derive-0.11.0 | check | full | -3.60% | 18.02x
pest_generator-2.1.3 | check | full | -3.53% | 17.63x
stdweb-derive-0.5.3 | check | full | -3.43% | 17.16x
enum-as-inner-0.4.0 | check | full | -3.41% | 17.05x
serde_derive-1.0.136 | check | full | -3.23% | 16.17x
clap_derive-3.1.7 | check | full | -3.22% | 16.10x
ref-cast-impl-1.0.6 | check | full | -3.12% | 15.61x
prost-derive-0.10.0 | check | full | -3.07% | 15.36x
wayland-scanner-0.29.4 | check | full | -3.05% | 15.23x
pyo3-macros-backend-0.16.3 | check | full | -3.02% | 15.09x

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 643e9f707ed4ca13a158b6e290b424e520809ca6 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 643e9f707ed4ca13a158b6e290b424e520809ca6 -->
<!-- homu: {"type":"Approved","sha":"643e9f707ed4ca13a158b6e290b424e520809ca6","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 643e9f707ed4ca13a158b6e290b424e520809ca6 with merge e35a533fb398e98b318502c2920f82e110aad604...
<!-- homu: {"type":"BuildStarted","head_sha":"643e9f707ed4ca13a158b6e290b424e520809ca6","merge_sha":"e35a533fb398e98b318502c2920f82e110aad604"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/6113161438?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/e35a533fb398e98b318502c2920f82e110aad604/checks/6113161438/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

lqd commented:
@bors retry "The hosted runner: GitHub Actions 5 lost communication with the server" on dist-aarch64-apple

bors commented:
:hourglass: Testing commit 643e9f707ed4ca13a158b6e290b424e520809ca6 with merge b04c5329e1e145fb2fb46c5a7e775638712b03aa...
<!-- homu: {"type":"BuildStarted","head_sha":"643e9f707ed4ca13a158b6e290b424e520809ca6","merge_sha":"b04c5329e1e145fb2fb46c5a7e775638712b03aa"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6117516108?check_suite_focus=true)
Approved by: petrochenkov
Pushing b04c5329e1e145fb2fb46c5a7e775638712b03aa to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6117516108?check_suite_focus=true"},"merge_sha":"b04c5329e1e145fb2fb46c5a7e775638712b03aa"} -->

rust-timer commented:
Finished benchmarking commit (b04c5329e1e145fb2fb46c5a7e775638712b03aa): [comparison url](https://perf.rust-lang.org/compare.html?start=1dec35a1b0df406da5d7cae55a7fa8d186a2b028&end=b04c5329e1e145fb2fb46c5a7e775638712b03aa).

**Summary**: 
- Primary benchmarks: 🎉 relevant improvements found
- Secondary benchmarks: 🎉 relevant improvements found

| | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:---:|:---:|:---:|:---:|:---:|:---:|
| count[^1] | 0 | 0 | 94 | 83 | 94 |
| mean[^2] | N/A | N/A | -0.9% | -2.4% | -0.9% |
| max | N/A | N/A | -2.4% | -17.2% | -2.4% |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
